tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs

```diff
@@ -7,254 +7,136 @@
-                    var separatorIndex = tag.IndexOf(KeyValueSeparator);
-
-                    // "_dd.p.a=b"
-                    //         ⬆   separator must be at index 7 or higher and before the end of string
-                    //  012345678
-                    if (separatorIndex > PropagatedTagPrefixLength &&
-                        separatorIndex < tag.Length - 1)
-                    {
-                        // TODO: implement something like StringSegment to avoid allocating new strings?
-                        var key = tag.Substring(0, separatorIndex);
-                        var value = tag.Substring(separatorIndex + 1);
-                        tagList.Add(new KeyValuePair<string, string>(key, value));
-                    }
-                    else
-                    {
-                        // skip invalid tag
-                        cacheOriginalHeader = false;
-                    }
-                }
-                else
-                {
-                    // skip invalid tag
-                    cacheOriginalHeader = false;
-                }
-            }
-
-            var traceTags = new TraceTagCollection(tagList);
+        var isPropagated = name.StartsWith(TagPropagation.PropagatedTagPrefix, StringComparison.Ordinal);
 
-            if (cacheOriginalHeader)
-            {
-                // we didn't skip any invalid tag, we can cache the original header string
-                traceTags._cachedPropagationHeader = propagationHeader;
-            }
-
-            return traceTags;
-        }
-
-        public void SetTag(string key, string? value)
+        lock (_listLock)
         {
-            if (key is null)
-            {
-                ThrowHelper.ThrowArgumentNullException(nameof(key));
-            }
+            _tags ??= new List<KeyValuePair<string, string>>(1);
 
-            lock (_tags)
+            if (_tags.Count > 0)
             {
-                bool tagsModified = false;
-
                 for (int i = 0; i < _tags.Count; i++)
                 {
-                    if (string.Equals(_tags[i].Key, key, StringComparison.Ordinal))
+                    if (string.Equals(_tags[i].Key, name, StringComparison.Ordinal))
                     {
                         if (value == null)
                         {
                             _tags.RemoveAt(i);
-                            tagsModified = true;
                         }
                         else if (!string.Equals(_tags[i].Value, value, StringComparison.Ordinal))
                         {
-                            _tags[i] = new KeyValuePair<string, string>(key, value);
-                            tagsModified = true;
-                        }
+                            _tags[i] = new(name, value);
 
-                        // clear the cached header if we make any changes to a distributed tag
-                        if (tagsModified && key.StartsWith(PropagatedTagPrefix, StringComparison.Ordinal))
```

Comment:
Not really related to this PR, but are tag names _meant_ to be case-sensitive? It seems kind of weird for that to be the case, given we normalise case down the line?

Reply:
Good catch! I made the tag names case-insensitive by switching to `StringComparison.OrdinalIgnoreCase`.